kernel/src/vm/vmar/mod.rs

```diff
@@ -694,18 +689,43 @@ impl Vmar_ {
         }
 
         // Clone mappings.
-        for (vm_mapping_base, vm_mapping) in &inner.vm_mappings {
-            let new_mapping = Arc::new(vm_mapping.new_fork(&new_vmar_)?);
-            new_vmar_
-                .inner
-                .lock()
-                .vm_mappings
-                .insert(*vm_mapping_base, new_mapping);
+        {
+            let new_vmspace = new_vmar_.vm_space();
+            let range = self.base..(self.base + self.size);
+            let mut new_cursor = new_vmspace.cursor_mut(&range).unwrap();
+            let cur_vmspace = self.vm_space();
+            let mut cur_cursor = cur_vmspace.cursor_mut(&range).unwrap();
+            for (vm_mapping_base, vm_mapping) in &inner.vm_mappings {
+                // Clone the `VmMapping` to the new VMAR.
+                let new_mapping = Arc::new(vm_mapping.new_fork(&new_vmar_)?);
+                new_vmar_
+                    .inner
+                    .lock()
+                    .vm_mappings
+                    .insert(*vm_mapping_base, new_mapping);
+
+                // Protect the mapping and copy to the new page table for COW.
+                cur_cursor.jump(*vm_mapping_base).unwrap();
+                new_cursor.jump(*vm_mapping_base).unwrap();
+                let mut op = |page: &mut PageProperty| {
+                    page.flags -= PageFlags::W;
+                };
+                new_cursor.copy_from(&mut cur_cursor, vm_mapping.map_size(), &mut op);
```

Comment:
Do we need to encapsulate this part into a function and perform a global TLB flush at the end?

Reply:
Indeed. Forgot to do so.